### PR TITLE
Fix issue #100, make work again with v0.5

### PR DESCRIPTION
**Pull request by ScottPJones**
This fixes deprecations due to removal of `call` overloading,
and also fixes a problem with handling subarrays,
due to https://github.com/JuliaLang/julia/pull/15071 getting rid of the unexported `index_generate` function.

### CODE REVIEW

**Comment by ScottPJones** [comment]
@andreasnoack @ElOceanografo Could somebody from JuliaStats review this and possibly merge this?
@mbauman Does my change based on your suggestion look correct for subarrays?  Thanks!!!

File: src/subarrayn.jl [hunk 1]
```diff
@@ -0,0 +1,25 @@
+const unsafe_getindex = Base.unsafe_getindex
+
+function Base.isnull{T,N,P<:NullableArray,IV,LD}(V::SubArray{T,N,P,IV,LD}, I::Int...)
+    @Base._inline_meta
```
Comment (mbauman) [inline src/subarrayn.jl]: Just use `@inline function...`. This private macro only exists for bootstrapping.

File: src/subarrayn.jl [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+end
+
+function Base.values{T,N,P<:NullableArray,IV,LD}(V::SubArray{T,N,P,IV,LD}, I::Int...)
+    @Base._inline_meta
+    @boundscheck checkbounds(V, I...)
+    @inbounds r = V.parent.values[Base.reindex(V, V.indexes, Base.to_indexes(I...))...]
+    r
+end
+
+@generated function anynull{T, N, U<:NullableArray}(S::SubArray{T, N, U})
```
Comment (mbauman) [inline src/subarrayn.jl]: You didn't make any changes to this fubction, did you? Let's not make unnecessary duplications across versions here. I'd just put the version switch in the existing SubArray file.

**Comment by mbauman** [comment]
There will be a performance regression for linear indexing with linear fast SubArrays here. That can be avoided with an additional method that uses the stride computation as in https://github.com/JuliaLang/julia/blob/57c53afba29a06493a2d993fb571a992d0ad733a/base/subarray.jl#L151.

File: src/subarrayn.jl [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const unsafe_getindex = Base.unsafe_getindex
+
+function Base.isnull{T,N,P<:NullableArray,IV,LD}(V::SubArray{T,N,P,IV,LD}, I::Int...)
+    @Base._inline_meta
+    @boundscheck checkbounds(V, I...)
+    @inbounds r = V.parent.isnull[Base.reindex(V, V.indexes, Base.to_indexes(I...))...]
```
Comment (mbauman) [inline src/subarrayn.jl]: Don't use `to_indexes`. They're already all Int.

**Comment by nalimilan** [comment]
Could you move the unrelated indentation changes to a separate commit?

File: src/subarrayn.jl [hunk 4]
```diff
@@ -0,0 +1,25 @@
+const unsafe_getindex = Base.unsafe_getindex
```
Comment (nalimilan) [inline src/subarrayn.jl]: This doesn't appear to be used anywhere.

Comment (ScottPJones) [inline src/subarrayn.jl]: That was just in the original code, so I left it in, didn't know if it was necessary. I'll remove it.

File: src/subarrayn.jl (hunk 2 above)
Comment (ScottPJones) [inline src/subarrayn.jl]: Originally I was running into problems with still getting because the parser was giving warnings, and the first way I tried to prevent that was to use separate files (it turned out that using `@compat` was enough to solve those problems).  Fixed now.

File: src/subarrayn.jl (hunk 3 above)
Comment (ScottPJones) [inline src/subarrayn.jl]: Turns out, to_indexes for some reason is necessary, if I don't use it, it gets an error, saying that it can't find a method of reindex that matches.  Any idea why?

File: src/subarrayn.jl (hunk 2 above)
Comment (ScottPJones) [inline src/subarrayn.jl]: It gives strange error messages under v0.4.5 (fine under v0.5) when I try to combine them:

``` julia
ERROR: LoadError: LoadError: wrong number of arguments
 in include at /Applications/Julia-0.4.5.app/Contents/Resources/julia/lib/julia/sys.dylib
 in include_from_node1 at /Applications/Julia-0.4.5.app/Contents/Resources/julia/lib/julia/sys.dylib
 in include at /Applications/Julia-0.4.5.app/Contents/Resources/julia/lib/julia/sys.dylib
 in include_from_node1 at /Applications/Julia-0.4.5.app/Contents/Resources/julia/lib/julia/sys.dylib
 in require at /Applications/Julia-0.4.5.app/Contents/Resources/julia/lib/julia/sys.dylib
while loading /Users/scott/.julia/v0.4/NullableArrays/src/subarray.jl, in expression starting on line 102
while loading /Users/scott/.julia/v0.4/NullableArrays/src/NullableArrays.jl, in expression starting on line 33
```

The error corresponds to the line after the `end` from the `if VERSION < v"0.5.0-dev+2718"` ... `else` ... `end`.
I have no idea what is causing this issue in v0.4.5, a bug in the compiler maybe?

File: src/subarrayn.jl (hunk 3 above)
Comment (mbauman) [inline src/subarrayn.jl]: Are you splatting `I`?  Don't splat `I`.

File: src/subarrayn.jl (hunk 2 above)
Comment (mbauman) [inline src/subarrayn.jl]: `@boundscheck` means something different in 0.4, where it takes two arguments.  That's probably the issue.  I'm not sure if Compat has this handled yet, but you should (eventually) be able to to `@compat @boundscheck …`.

Comment (ScottPJones) [inline src/subarrayn.jl]: OK, for the moment, I put the new stuff into `subarraynew.jl`, and included it in `subarray.jl`.  That's the only way I could think of to completely avoid parsing (and executing the macros) the code for v0.5.  Is there a better technique that is recommended?  Thanks again for the review!

File: src/subarrayn.jl (hunk 4 above)
Comment (ScottPJones) [inline src/subarrayn.jl]: Turns out it is necessary for the v0.4 code.

**Comment by ScottPJones** [comment]
@mbauman @nalimilan I think this is all set to go now.  Thanks to both of you for the review (learned some neat stuff about the new subarray changes to boot!)

File: src/subarraynew.jl [hunk 5]
```diff
@@ -0,0 +1,36 @@
+typealias NullableSubArray{T,N,P<:NullableArray,IV,LD} SubArray{T,N,P,IV,LD}
+
+@inline function Base.isnull(V::NullableSubArray, I::Real...)
+    @boundscheck checkbounds(V, I...)
+    @inbounds return V.parent.isnull[Base.reindex(V, V.indexes, Base.to_indexes(I...))...]
```
Comment (mbauman) [inline src/subarraynew.jl]: Keep the signature the same between 0.4 and 0.5, and don't use `to_indexes`.  Widening the signature is an unrelated change from what's required here, and it introduces a feature difference between code that's meant to be functionally the same.

File: src/subarrayn.jl (hunk 2 above)
Comment (mbauman) [inline src/subarrayn.jl]: Ah, yes, I see that hasn't been done yet (https://github.com/JuliaLang/Compat.jl/issues/182).  This seems to be an okay solution for now.

**Comment by mbauman** [comment]
Besides https://github.com/JuliaStats/NullableArrays.jl/pull/101#discussion_r58319542 the SubArray commit looks good to me.  I didn't review the other commits.

File: src/subarraynew.jl (hunk 5 above)
Comment (ScottPJones) [inline src/subarraynew.jl]: It didn't work without `to_indexes`.  I removed them, and then got an method error about `reindex`.

Comment (mbauman) [inline src/subarraynew.jl]: What did you try, and what error did you get?  Don't splat `I`.  Just `V.parent.isnull[Base.reindex(V, V.indexes, I)...]`.

Comment (ScottPJones) [inline src/subarraynew.jl]: Never mind - the problem was an issue with splatting, I needed to remove both the to_index call _and_ the splatting.

**Comment by ScottPJones** [comment]
OK, done now, fixed all of the subarray issues, the other issue was simply using `@compat` in 3 places to handle the deprecation of `call` overloading.

File: src/subarraynew.jl (hunk 5 above)
Comment (El-Zurdo-Sordo) [inline src/subarraynew.jl]: ...and I didn't see your note until after I'd finally figured it out myself. duh!

**Comment by mbauman** [comment]
LGTM.  Thanks @ScottPJones.

File: src/subarraynew.jl (hunk 5 above)
Comment (nalimilan) [inline src/subarraynew.jl]: Why do you introduce these `to_index` calls in one commit, and remove them in the next one?

Comment (El-Zurdo-Sordo) [inline src/subarraynew.jl]: @nalimilan They were necessary to have the subscripts take `Real`, which would be more consistent with the code in `Base/subarray.jl`, however, I removed them based on @mbauman review comment above.
I do feel that it's incorrect to have these act differently than plain SubArrays, however, that could be fixed in a separate PR if people agree.

Comment (ScottPJones) [inline src/subarraynew.jl]: @nalimilan They were necessary to have the subscripts take Real, which would be more consistent with the code in Base/subarray.jl, however, I removed them based on @mbauman review comment above.
I do feel that it's incorrect to have these act differently than plain SubArrays, however, that could be fixed in a separate PR if people agree.

Comment (kmsquire) [inline src/subarraynew.jl]: Presumably the commits will be squashed, removing the the addiction and
subtraction of that feature?  I believe this is what @naliman was referring
to.

Comment (ScottPJones) [inline src/subarraynew.jl]: @kmsquire I was asked to keep the fixing of the indentation as a separate commit, and I think the fixes related to adding Compat should probably be in a separate commit from the fixes related to JuliaLang/julia#15071, but I'll squash out that commit.  Does that sound reasonable?

Comment (kmsquire) [inline src/subarraynew.jl]: Perfect

Comment (nalimilan) [inline src/subarraynew.jl]: Yes, that was what I meant.

**Comment by ScottPJones** [comment]
Bump!  Anybody can merge this fix around?  Thanks!

**Comment by davidagold** [comment]
@ScottPJones Thank you very much for this =) I have just one request -- could you please change the filename`subarraynew.jl` to something more specific, like `subarray0_5.jl`? Does that seem like a reasonable request?

**Comment by ScottPJones** [comment]
@davidagold Very reasonable! (I should have thought of that myself! 😳)  Thank _you_ for the nice package!
